File: euchre/infoset.py
"""Information sets and observation encoding.

Two views of a state from one player's perspective:

* :func:`infoset_key` — a canonical string identifying the information set,
  used to key tabular CFR regret/strategy tables. Two histories a player
  cannot tell apart must produce the same key.
* :func:`observation_tensor` — a fixed-length float vector for neural
  networks. Positions are encoded *relative* to the acting player so the same
  strategic situation looks identical from any seat.

The hidden information (other players' hands, the kitty) is never included:
both views are strictly what ``player`` can legally observe.
"""

from __future__ import annotations

from typing import List

import numpy as np

from .cards import (
    Card, Suit, Rank, RANKS, SUITS, NUM_CARDS, _RANK_INDEX,
    same_color_suit, effective_suit, is_trump, is_left_bower, is_right_bower,
)
from .game import EuchreState, Phase, team_of


def _rel(seat: int, me: int) -> int:
    """Seat expressed relative to the acting player (0 = me, 1 = left, ...)."""
    return (seat - me) % 4


def infoset_key(state: EuchreState, player: int) -> str:
    """Canonical string for the information set of ``player`` at ``state``."""
    my_score, their_score = (
        (state.team0_score, state.team1_score) if team_of(player) == 0
        else (state.team1_score, state.team0_score))
    parts: List[str] = [f"ph{state.phase.value}", f"d{_rel(state.dealer, player)}",
                        f"sc{my_score},{their_score}"]

    # Private hand (sorted by id for canonical order).
    hand_ids = sorted(c.id for c in state.hands[player])
    parts.append("h" + ",".join(map(str, hand_ids)))

    # Up-card is public during bidding.
    if state.phase in (Phase.BID_ROUND_1, Phase.BID_ROUND_2,
                       Phase.DEALER_DISCARD) and state.up_card is not None:
        parts.append(f"u{state.up_card.id}")
    if state.turned_down is not None:
        parts.append(f"td{int(state.turned_down)}")
    parts.append(f"b{state.bids_seen}")

    # Trump / maker once known.
    if state.trump is not None:
        parts.append(f"t{int(state.trump)}")
        parts.append(f"m{_rel(state.maker, player)}")
        parts.append(f"a{int(state.alone)}")

    # Public play history: completed tricks (winner + ordered plays) and the
    # current trick, all with relative seats.
    for winner, plays in state.completed_tricks:
        seq = ";".join(f"{_rel(pl, player)}:{c.id}" for pl, c in plays)
        parts.append(f"T{_rel(winner, player)}|{seq}")
    cur = ";".join(f"{_rel(pl, player)}:{c.id}" for pl, c in state.current_trick)
    parts.append(f"C{cur}")
    parts.append(f"w{state.tricks_won[0]},{state.tricks_won[1]}")
    return "/".join(parts)


# --- Fixed-length observation vector (suit-agnostic / trump-relative) --------
#
# Every suit-bearing feature is encoded by its ROLE relative to a reference
# suit R (= trump once set, else the up-card's suit during bidding), never by
# absolute suit identity. The two off-color ("green") suits share the single
# "green" role tag, so a network built on this encoding cannot develop a
# per-absolute-suit preference: relabeling the two greens (or any color-
# preserving suit permutation) permutes the observation's per-suit blocks
# without otherwise changing it. See docs/rebel_design.md and the plan.
#
# Layout: [ global block | 4 per-suit blocks (absolute slot order, role-
# relative contents) | 24 per-card feature blocks (card-id order) ]. The
# per-suit blocks stay in ABSOLUTE order so a shared network tower maps each
# to its own Call/Play output index by construction; only their *contents*
# are role-relative, which is what makes the tower suit-agnostic.

_PHASES = [Phase.BID_ROUND_1, Phase.BID_ROUND_2, Phase.DEALER_DISCARD,
           Phase.PLAY, Phase.TERMINAL]

# Race-to-target match score. A hand's starting score is always < this for
# both teams (the match ends, and no further hand is dealt, the instant a
# team reaches it), so team score / MATCH_TARGET always lands in [0, 1).
# rebel/match_equity.py imports this so the equity table and the observation
# encoding always agree on what "match" means.
MATCH_TARGET = 10

# Roles of a suit relative to the reference suit R.
_ROLE_REF, _ROLE_NEXT, _ROLE_GREEN = 0, 1, 2
N_ROLES = 3

# The 7 trump-rank "slots" for the as-if-this-suit-were-trump holdings, in
# strength order: right bower, left bower, then A/K/Q/10/9 of the suit.
_TRUMP_HOLDING_RANKS = [Rank.ACE, Rank.KING, Rank.QUEEN, Rank.TEN, Rank.NINE]

# --- segment dims ---
_GLOBAL = (
    len(_PHASES)   # phase one-hot
    + 4            # dealer, relative seat
    + 5            # maker: none + 4 relative seats
    + 1            # alone
    + 2            # tricks won: mine, theirs (/5)
    + 1            # am I on lead this trick?
    + 1            # bids_seen (/7)
    + 1            # up-card visible?
    + 6            # up-card rank one-hot (belongs to the reference suit)
    + 4            # led-card role this trick: none/ref/next/green
    + 2            # match score: mine, theirs (/MATCH_TARGET)
)  # = 32

_SUIT_BLOCK = (
    N_ROLES        # role one-hot (ref/next/green)
    + 1            # is the turned-down suit (illegal to call in round 2)
    + 1            # is the actual trump suit
    + 7            # as-if-trump holdings: RB, LB, A, K, Q, 10, 9
    + 1            # as-if-trump count (/5)
    + 6            # effective plain-suit holdings under actual trump (by rank)
    + 1            # void in this effective suit
    + 4            # cards of this effective suit played, per relative seat (/5)
    + 1            # cards of this suit seen so far (/6)
)  # = 25
NUM_SUITS = 4

_CARD_FEAT = (
    6              # rank one-hot
    + 1            # in my hand
    + 1            # is right bower (actual trump)
    + 1            # is left bower (actual trump)
    + 1            # is effective trump (actual trump)
    + 1            # has been played this hand
)  # = 11

GLOBAL_OFF = 0
SUIT_OFF = _GLOBAL
SUIT_BLOCK_DIM = _SUIT_BLOCK
GLOBAL_DIM = _GLOBAL
CARD_FEAT_DIM = _CARD_FEAT
CARD_OFF = _GLOBAL + NUM_SUITS * _SUIT_BLOCK
OBS_SIZE = _GLOBAL + NUM_SUITS * _SUIT_BLOCK + NUM_CARDS * _CARD_FEAT


def _reference_suit(state: EuchreState):
    """The suit everything is encoded relative to: trump once set, else the
    up-card's suit during bidding. None only in DEAL/TERMINAL (no decision)."""
    if state.trump is not None:
        return state.trump
    if state.up_card is not None:
        return state.up_card.suit
    return None


def _role_of(suit: Suit, ref) -> int:
    if ref is None:
        return _ROLE_GREEN  # no reference (DEAL/TERMINAL); role is unused
    if suit == ref:
        return _ROLE_REF
    if suit == same_color_suit(ref):
        return _ROLE_NEXT
    return _ROLE_GREEN


def observation_tensor(state: EuchreState, player: int) -> np.ndarray:
    v = np.zeros(OBS_SIZE, dtype=np.float32)
    hand = state.hands[player]
    hand_set = set(hand)
    trump = state.trump
    ref = _reference_suit(state)
    show_up = (state.up_card is not None and trump is None)

    # ---- global block ----
    o = GLOBAL_OFF
    v[o + _PHASES.index(state.phase)] = 1.0
    o += len(_PHASES)
    v[o + _rel(state.dealer, player)] = 1.0
    o += 4
    v[o + (0 if state.maker is None else _rel(state.maker, player) + 1)] = 1.0
    o += 5
    v[o] = 1.0 if state.alone else 0.0
    o += 1
    my_team = player % 2
    v[o] = state.tricks_won[my_team] / 5.0
    v[o + 1] = state.tricks_won[1 - my_team] / 5.0
    o += 2
    my_score = state.team0_score if my_team == 0 else state.team1_score
    their_score = state.team1_score if my_team == 0 else state.team0_score
    v[o] = my_score / MATCH_TARGET
    v[o + 1] = their_score / MATCH_TARGET
    o += 2
    v[o] = 1.0 if (state.phase == Phase.PLAY and not state.current_trick) else 0.0
    o += 1
    v[o] = state.bids_seen / 7.0
    o += 1
    v[o] = 1.0 if show_up else 0.0
    o += 1
    if show_up:
        v[o + _RANK_INDEX[state.up_card.rank]] = 1.0
    o += 6
    # led-card role this trick: index 0 = no led card, else role+1
    if state.current_trick:
        led_role = _role_of(effective_suit(state.current_trick[0][1], trump), ref)
        v[o + led_role + 1] = 1.0
    else:
        v[o] = 1.0
    o += 4
    assert o == SUIT_OFF

    # ---- per-suit blocks (absolute slot, role-relative contents) ----
    # Precompute public play info once.
    played_cards = []  # (rel_seat, card)
    for _winner, plays in state.completed_tricks:
        for seat, c in plays:
            played_cards.append((_rel(seat, player), c))
    for seat, c in state.current_trick:
        played_cards.append((_rel(seat, player), c))

    for s in SUITS:
        o = SUIT_OFF + int(s) * _SUIT_BLOCK
        v[o + _role_of(s, ref)] = 1.0
        o += N_ROLES
        v[o] = 1.0 if state.turned_down == s else 0.0
        o += 1
        v[o] = 1.0 if trump == s else 0.0
        o += 1
        # as-if-s-were-trump holdings: right bower, left bower, A, K, Q, 10, 9
        v[o] = 1.0 if Card(s, Rank.JACK) in hand_set else 0.0
        v[o + 1] = 1.0 if Card(same_color_suit(s), Rank.JACK) in hand_set else 0.0
        for i, r in enumerate(_TRUMP_HOLDING_RANKS):
            v[o + 2 + i] = 1.0 if Card(s, r) in hand_set else 0.0
        o += 7
        as_if_trump = sum(1 for c in hand if is_trump(c, s))
        v[o] = as_if_trump / 5.0
        o += 1
        # effective plain-suit holdings under ACTUAL trump (by rank)
        eff_here = [c for c in hand if effective_suit(c, trump) == s]
        for c in eff_here:
            v[o + _RANK_INDEX[c.rank]] = 1.0
        o += 6
        v[o] = 1.0 if not eff_here else 0.0
        o += 1
        # cards of this effective suit played, per relative seat (/5)
        for rel_seat, c in played_cards:
            if effective_suit(c, trump) == s:
                v[o + rel_seat] += 1.0 / 5.0
        o += 4
        # cards of this (effective) suit seen so far. Divisor 7 because the
        # effective trump suit spans 7 cards (its own 6 + the left bower); all
        # other effective suits are smaller, so this keeps the feature in
        # [0, 1] for every suit/role.
        seen = len(eff_here) + sum(1 for _rs, c in played_cards
                                   if effective_suit(c, trump) == s)
        if show_up and effective_suit(state.up_card, trump) == s:
            seen += 1
        v[o] = seen / 7.0
        o += 1

    # ---- per-card feature blocks (card-id order) ----
    played_set = {c for _rs, c in played_cards}
    for cid in range(NUM_CARDS):
        c = Card.from_id(cid)
        o = CARD_OFF + cid * _CARD_FEAT
        v[o + _RANK_INDEX[c.rank]] = 1.0
        o += 6
        v[o] = 1.0 if c in hand_set else 0.0
        o += 1
        v[o] = 1.0 if (trump is not None and is_right_bower(c, trump)) else 0.0
        o += 1
        v[o] = 1.0 if (trump is not None and is_left_bower(c, trump)) else 0.0
        o += 1
        v[o] = 1.0 if (trump is not None and is_trump(c, trump)) else 0.0
        o += 1
        v[o] = 1.0 if c in played_set else 0.0
        o += 1

    return v

File: cpp/infoset.cpp
#include "infoset.h"

#include <algorithm>
#include <cstring>
#include <sstream>

namespace mceuchre {

namespace {

inline int rel(int seat, int me) { return ((seat - me) % 4 + 4) % 4; }

// Index into the 5-entry phase list used for the observation's phase one-hot
// (mirrors euchre/infoset.py's _PHASES -- Deal is deliberately absent;
// observation_tensor is never called on a Deal-phase state).
int phase_obs_index(Phase p) {
    switch (p) {
        case Phase::BidRound1: return 0;
        case Phase::BidRound2: return 1;
        case Phase::DealerDiscard: return 2;
        case Phase::Play: return 3;
        case Phase::Terminal: return 4;
        default: throw std::runtime_error("observation_tensor called on Deal phase");
    }
}

// Matches Python's Phase(enum.auto()) numbering exactly (declaration order,
// 1-based): Deal=1, BidRound1=2, BidRound2=3, DealerDiscard=4, Play=5,
// Terminal=6. infoset_key must byte-match the Python string, which embeds
// this value directly.
int phase_value(Phase p) {
    switch (p) {
        case Phase::Deal: return 1;
        case Phase::BidRound1: return 2;
        case Phase::BidRound2: return 3;
        case Phase::DealerDiscard: return 4;
        case Phase::Play: return 5;
        case Phase::Terminal: return 6;
    }
    throw std::runtime_error("unreachable");
}

// The suit everything is encoded relative to: trump once set, else the
// up-card's suit during bidding. -1 only in Deal/Terminal (no decision).
int reference_suit(const EuchreState& s) {
    if (s.trump.has_value()) return *s.trump;
    if (s.up_card.has_value()) return card_suit(*s.up_card);
    return -1;
}

int role_of(int suit, int ref) {
    if (ref < 0) return ROLE_GREEN;  // no reference; role is unused
    if (suit == ref) return ROLE_REF;
    if (suit == same_color_suit(ref)) return ROLE_NEXT;
    return ROLE_GREEN;
}

// _TRUMP_HOLDING_RANKS = [ACE, KING, QUEEN, TEN, NINE] -> rank indices
// (0=9,1=10,2=J,3=Q,4=K,5=A in this engine's numbering).
constexpr int TRUMP_HOLDING_RANK_IDX[5] = {5, 4, 3, 1, 0};

}  // namespace

void observation_tensor(const EuchreState& state, int player, float* out) {
    std::memset(out, 0, sizeof(float) * OBS_SIZE);
    Hand hand = state.hands[player];
    int trump = state.trump.has_value() ? *state.trump : -1;
    int ref = reference_suit(state);
    bool show_up = state.up_card.has_value() && !state.trump.has_value();

    // ---- global block ----
    int o = GLOBAL_OFF;
    out[o + phase_obs_index(state.phase)] = 1.0f;
    o += 5;
    out[o + rel(state.dealer, player)] = 1.0f;
    o += 4;
    out[o + (state.maker.has_value() ? rel(*state.maker, player) + 1 : 0)] = 1.0f;
    o += 5;
    out[o] = state.alone ? 1.0f : 0.0f;
    o += 1;
    int my_team = player % 2;
    out[o] = state.tricks_won[my_team] / 5.0f;
    out[o + 1] = state.tricks_won[1 - my_team] / 5.0f;
    o += 2;
    int my_score = my_team == 0 ? state.team0_score : state.team1_score;
    int their_score = my_team == 0 ? state.team1_score : state.team0_score;
    out[o] = static_cast<float>(my_score) / MATCH_TARGET;
    out[o + 1] = static_cast<float>(their_score) / MATCH_TARGET;
    o += 2;
    out[o] = (state.phase == Phase::Play && state.current_trick.empty()) ? 1.0f : 0.0f;
    o += 1;
    out[o] = static_cast<float>(state.bids_seen) / 7.0f;
    o += 1;
    out[o] = show_up ? 1.0f : 0.0f;
    o += 1;
    if (show_up) out[o + card_rank(*state.up_card)] = 1.0f;
    o += 6;
    if (!state.current_trick.empty()) {
        int led_role = role_of(effective_suit(state.current_trick[0].card, trump), ref);
        out[o + led_role + 1] = 1.0f;
    } else {
        out[o] = 1.0f;
    }
    o += 4;
    // assert(o == SUIT_OFF);

    // ---- per-suit blocks ----
    // Precompute public play info once: (rel_seat, card) for every card
    // played so far (completed tricks + current trick).
    std::vector<std::pair<int, CardId>> played_cards;
    for (const auto& t : state.completed_tricks)
        for (const auto& p : t.plays) played_cards.emplace_back(rel(p.player, player), p.card);
    for (const auto& p : state.current_trick)
        played_cards.emplace_back(rel(p.player, player), p.card);

    for (int s = 0; s < NUM_SUITS; ++s) {
        o = SUIT_OFF + s * SUIT_BLOCK_DIM;
        out[o + role_of(s, ref)] = 1.0f;
        o += N_ROLES;
        out[o] = (state.turned_down.has_value() && *state.turned_down == s) ? 1.0f : 0.0f;
        o += 1;
        out[o] = (trump == s) ? 1.0f : 0.0f;
        o += 1;
        // as-if-s-were-trump holdings: right bower, left bower, A, K, Q, 10, 9
        out[o] = hand_has(hand, make_card(s, 2)) ? 1.0f : 0.0f;              // right bower (Jack of s)
        out[o + 1] = hand_has(hand, make_card(same_color_suit(s), 2)) ? 1.0f : 0.0f;  // left bower
        for (int i = 0; i < 5; ++i)
            out[o + 2 + i] = hand_has(hand, make_card(s, TRUMP_HOLDING_RANK_IDX[i])) ? 1.0f : 0.0f;
        o += 7;
        int as_if_trump = 0;
        for (CardId c : hand_to_vector(hand)) if (is_trump(c, s)) as_if_trump++;
        out[o] = static_cast<float>(as_if_trump) / 5.0f;
        o += 1;
        // effective plain-suit holdings under ACTUAL trump (by rank)
        std::vector<CardId> eff_here;
        for (CardId c : hand_to_vector(hand))
            if (effective_suit(c, trump) == s) eff_here.push_back(c);
        for (CardId c : eff_here) out[o + card_rank(c)] = 1.0f;
        o += 6;
        out[o] = eff_here.empty() ? 1.0f : 0.0f;
        o += 1;
        // cards of this effective suit played, per relative seat (/5)
        for (const auto& [rel_seat, c] : played_cards)
            if (effective_suit(c, trump) == s) out[o + rel_seat] += 1.0f / 5.0f;
        o += 4;
        // cards of this (effective) suit seen so far. Divisor 7: effective
        // trump suit spans 7 cards (its own 6 + the left bower).
        int seen = static_cast<int>(eff_here.size());
        for (const auto& [_rs, c] : played_cards)
            if (effective_suit(c, trump) == s) seen++;
        if (show_up && effective_suit(*state.up_card, trump) == s) seen++;
        out[o] = static_cast<float>(seen) / 7.0f;
        o += 1;
    }

    // ---- per-card feature blocks (card-id order) ----
    std::vector<bool> played_set(NUM_CARDS, false);
    for (const auto& [_rs, c] : played_cards) played_set[c] = true;
    for (CardId cid = 0; cid < NUM_CARDS; ++cid) {
        o = CARD_OFF + cid * CARD_FEAT_DIM;
        out[o + card_rank(cid)] = 1.0f;
        o += 6;
        out[o] = hand_has(hand, cid) ? 1.0f : 0.0f;
        o += 1;
        out[o] = (state.trump.has_value() && is_right_bower(cid, *state.trump)) ? 1.0f : 0.0f;
        o += 1;
        out[o] = (state.trump.has_value() && is_left_bower(cid, *state.trump)) ? 1.0f : 0.0f;
        o += 1;
        out[o] = (state.trump.has_value() && is_trump(cid, *state.trump)) ? 1.0f : 0.0f;
        o += 1;
        out[o] = played_set[cid] ? 1.0f : 0.0f;
        o += 1;
    }
}

std::string infoset_key(const EuchreState& state, int player) {
    int my_score, their_score;
    if (team_of(player) == 0) {
        my_score = state.team0_score; their_score = state.team1_score;
    } else {
        my_score = state.team1_score; their_score = state.team0_score;
    }

    std::ostringstream os;
    os << "ph" << phase_value(state.phase)
       << "/d" << rel(state.dealer, player)
       << "/sc" << my_score << "," << their_score;

    // Private hand (sorted by id for canonical order).
    std::vector<CardId> hand_ids = hand_to_vector(state.hands[player]);
    os << "/h";
    for (size_t i = 0; i < hand_ids.size(); ++i) {
        if (i) os << ",";
        os << static_cast<int>(hand_ids[i]);
    }

    // Up-card is public during bidding.
    if ((state.phase == Phase::BidRound1 || state.phase == Phase::BidRound2 ||
        state.phase == Phase::DealerDiscard) && state.up_card.has_value()) {
        os << "/u" << static_cast<int>(*state.up_card);
    }
    if (state.turned_down.has_value()) os << "/td" << static_cast<int>(*state.turned_down);
    os << "/b" << state.bids_seen;

    if (state.trump.has_value()) {
        os << "/t" << static_cast<int>(*state.trump)
           << "/m" << rel(*state.maker, player)
           << "/a" << (state.alone ? 1 : 0);
    }

    for (const auto& t : state.completed_tricks) {
        os << "/T" << rel(t.winner, player) << "|";
        for (size_t i = 0; i < t.plays.size(); ++i) {
            if (i) os << ";";
            os << rel(t.plays[i].player, player) << ":" << static_cast<int>(t.plays[i].card);
        }
    }
    os << "/C";
    for (size_t i = 0; i < state.current_trick.size(); ++i) {
        if (i) os << ";";
        os << rel(state.current_trick[i].player, player) << ":"
           << static_cast<int>(state.current_trick[i].card);
    }
    os << "/w" << state.tricks_won[0] << "," << state.tricks_won[1];
    return os.str();
}

void legal_mask(const EuchreState& state, bool* out) {
    std::fill(out, out + NUM_ACTIONS, false);
    for (const auto& a : state.legal_actions()) {
        out[a.index()] = true;
    }
}

}  // namespace mceuchre
